// @builtioflow/models/activities.js
const mongoose = require('mongoose')
const Schema = mongoose.Schema

const ActivitySchema = new Schema({
  uid: { type: String },
  version: { type: String },
  name: { type: String },
  label: { type: String },
  icon: { type: String },
  category: { type: String, default: 'service' },
  input: { type: Schema.Types.Mixed },
  output: { type: Schema.Types.Mixed },
  tags: { type: [] },
  published: { type: Boolean, default: true },
  hide_settings: { type: Boolean, default: false },
  act_type: { type: String },
  created_at: { type: Date, default: new Date() },
  updated_at: { type: Date, default: new Date() },
  deleted_at: { type: Date, default: null },
  usage: { type: Schema.Types.Mixed },
  validation: { type: Schema.Types.Mixed },
  default_value: { type: String },
  development_phase: { type: String },
  marketing_category: { type: [] },
  express_enabled: { type: Boolean },
  advanced_enabled: { type: Boolean }
})

mongoose.model('Activity', ActivitySchema)
module.exports = ['Activity', ActivitySchema]
