// @builtioflow/models/bills.js
const mongoose = require('mongoose')
const crypto = require('crypto')
const Schema = mongoose.Schema


const BillSchema = new Schema({
	uid: { type: String },
	flow_name: { type: String },
	flow_uid: { type: String },
	user_uid: { type: String },
	project_uid: { type: String },
	env_uid: { type: String },
	org_uid: { type: String },
	start_time: { type: Date },
	stop_time: { type: Date },
	duration: { type: Number },
	activity_name: { type: String },
	queue: { label: { type: String }, value: { type: String } },
	credits: { type: Number, default: 0 },
	timeout: { type: Number, default: 0 },
	billable: { type: Boolean },
	created_at: { type: Date, default: Date.now },
	updated_at: { type: Date, default: Date.now },
	deleted_at: { type: Date, default: null },
	result: { type: String },
	executed_by: { type: String },
    manual_run: { type: Boolean, default: false }
})

BillSchema.pre('save', function (next) {
	if (!this || !this.uid) {
		this.uid = 'bid' + crypto.randomBytes(18).toString('hex')
	}
	next()
})

mongoose.model('Bill', BillSchema)
module.exports = ['Bill', BillSchema]
