// @builtioflow/models/consumer_plan.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const ConsumerPlanSchema = new Schema({
    uid: { type: String },
    name: { type: String },
    plan_code: { type: String },
    description: { type: String },
    consumer_count: { type: Number, default: 0},
    type: { type: String, default: "group"},
    transactions: { type: Number},
    used    : { type : Number, default: 0},
    //integrations: [{ type: Schema.Types.ObjectId, ref: 'SynergyEnabledAuth' }],
    users:[{ type: Schema.Types.ObjectId, ref: 'User' }],
    default: {type: Boolean,default: false},
    // price: { type: String },
    // billing_cycle: { type: String, default: "month" },    default: { type: Boolean ,default: false},
    organization:{ type : Schema.Types.ObjectId, ref: 'Organization' },
    deleted_at: { type: Date, default: null },
    created_at: { type: Date },
    updated_at: { type: Date}
});

mongoose.model('ConsumerPlan', ConsumerPlanSchema)
module.exports = ['ConsumerPlan', ConsumerPlanSchema]
