// @builtioflow/models/flowtests.js
var mongoose = require('mongoose');
var crypto = require('crypto'),
    Schema = mongoose.Schema;


var FlowTestSchema = new Schema({
    uid: { type: String },
    flow_uid: { type: String },
    sid: { type: String },
    message: { type: String },
    type: { type: String, default: 'webhook' },
    time_stamp: { type: Number, default: Date.now },
    activity_name: { type: String, default: 'start' },
    activity_id: { type: String, default: 'start' },
    activity_label: { type: String, default: 'Start' },
    version: { type: String, default: 'v1' },
    user_uid: { type: String },
    created_at: { type: Date, default: Date.now },
    project_uid: { type: String, default: null },
    org_uid: { type: String, default: null },
    env_uid: { type: String, default: null }
});

FlowTestSchema.pre('save', function (next) {
    if (!this.uid) {
        this.uid = 'ft' + crypto.randomBytes(22).toString('hex');
    }
    next();
});

//mongooseLogs.model('FlowTest', FlowTestSchema, 'logs');
mongoose.model('FlowTest', FlowTestSchema);

module.exports = ['FlowTest', FlowTestSchema]
