// @builtioflow/models/flows.js
const mongoose = require('mongoose')
const Schema = mongoose.Schema
const crypto = require('crypto')

const FlowSchema = new Schema({
  uid: { type: String },
  is_express: { type: Boolean, default: false },
  lower_name: { type: String },
  app_user_object_uid: { type: String },
  required: { type: String },
  required_data: { type: String },
  author: [],
  resources: {
    auth: [{ type: String }],
    connection: [{ type: String }],
    enterpriseagent: [{ type: String }]
  },
  can: { type: Schema.Types.Mixed },
  description: { type: String },
  is_draft: { type: Boolean, default: false },
  name: { type: String },
  orig_author: [],
  is_published: { type: Boolean, default: false },
  template_id: { type: String },
  is_scheduled: { type: Boolean, default: false },
  schedule_id: { type: Number, default: 0 },
  scheduled: { type: Schema.Types.Mixed },
  schedule_lock: { type: Boolean, default: false },
  settings: { type: Schema.Types.Mixed },
  sid: { type: String },
  environment: { type: Schema.Types.Mixed },
  version: { type: Number },
  is_webhook_active: { type: Boolean },
  activities: { type: String },
  last_run: {
    credits: { type: Number, default: 0 },
    queue: { type: String },
    _date: { type: Date },
    duration: { type: Number, default: 0 }
  },
  total_run_time: { type: Number, default: 0 },
  canvas_settings: { type: Schema.Types.Mixed },
  total_run_count: { type: Number, default: 0 },
  failed_run_count: { type: Number, default: 0 },
  current_month: { type: Number, default: 0 },
  current_month_run_time: { type: Number, default: 0 },
  errors_obj: [],
  tags: [],
  activity_icons: [],
  trigger_icons: [],
  return_webhook: { type: Boolean },
  debugger: { type: Boolean, default: false },
  deleted_at: { type: Date, default: null },
  flow_copy_counter: { type: Number, default: 0 },
  is_live: { type: Boolean, default: false },
  container_size: { type: Number, default: 256 },
  trigger: {
    icon: { type: String },
    active: { type: Boolean, default: false },
    handlers: []
  },
  is_empty: { type: Boolean, default: false },
  live_visualization: { type: Boolean, default: false },
  flow_active: { type: Boolean, default: true },
  previously_active_states: {
    manual_run: { type: Boolean, default: false },
    webhook: { type: Boolean, default: false },
    trigger: { type: Boolean, default: false }
  },
  express_trigger_data: [],
  show_notes: { type: Boolean, default: false },
  flow_type: { type: String }
}, { 
  timestamps: { 
    createdAt: 'created_at', 
    updatedAt: 'updated_at' 
  } 
})

mongoose.model('Flow', FlowSchema)
module.exports = ['Flow', FlowSchema]
