// @builtioflow/models/logs.js
const mongoose = require('mongoose')
const crypto = require('crypto')
const Schema = mongoose.Schema


const LogSchema = new Schema({
  uid: { type: String },
  flow_uid: { type: String },
  bill_uid: { type: String },
  message: { type: String },
  message_exp: { type: String },
  type: { type: String },
  counter: { type: Number },
  total: { type: Number },
  ingroup: { type: String },
  time_stamp: { type: Number },
  activity_name: { type: String },
  activity_id: { type: String },
  activity_label: { type: String },
  version: { type: String },
  user_uid: { type: String },
  error: { type: Boolean, default: false },
  created_at: { type: Date, default: Date.now },
  updated_at: { type: Date, default: Date.now },
  debugger: { type: String },
  project_uid: { type: String },
  org_uid: { type: String, default: null },
  env_uid: { type: String, default: null }
})

LogSchema.pre('save', function (next) {
  if (!this || !this.uid) {
    this.uid = 'log' + crypto.randomBytes(18).toString('hex')
  }
  next()
});

(mongooseLogs || mongoose).model('Log', LogSchema)
module.exports = ['Log', LogSchema]
