// @builtioflow/models/newflows.js
const mongoose = require('mongoose')
const Schema = mongoose.Schema
const crypto = require('crypto')

let NewFlowSchema = new Schema({
  uid: { type: String },
  lower_name: { type: String },
  owner_uid: { type: String },
  created_by: { type: String },
  updated_by: { type: String },
  locked_by: { type: String, default: null },
  org_uid: { type: String },
  env_uid: { type: String },
  project_uid: { type: String },
  author: [],
  description: { type: String },
  is_draft: { type: Boolean, default: false },
  debugger: { type: Boolean, default: false },
  name: { type: String },
  orig_author: [],
  template_id: { type: String },
  settings: { type: Schema.Types.Mixed },
  sid: { type: String },
  environment: { type: Schema.Types.Mixed },
  version: { type: Number },
  activities: { type: String },
  last_run: {
    credits: { type: Number, default: 0 },
    queue: { type: String },
    _date: { type: Date },
    duration: { type: Number, default: 0 }
  },
  total_run_time: { type: Number, default: 0 },
  canvas_settings: { type: Schema.Types.Mixed },
  total_run_count: { type: Number, default: 0 },
  failed_run_count: { type: Number, default: 0 },
  current_month: { type: Number, default: 0 },
  current_month_run_time: { type: Number, default: 0 },
  container_size: { type: Number, default: 256 },
  errors_obj: [],
  tags: [],
  activity_icons: [],
  deleted_at: { type: Date, default: null },
  flow_copy_counter: { type: Number, default: 0 },
  trigger: {
    icon: { type: String },
    active: { type: Boolean, default: false },
    handlers: []
  },
  active: { type: Boolean, default: false },
  active_config: { type: String },
  live_visualization: { type: Boolean, default: false },
  express_trigger_data: [],
  show_notes: { type: Boolean, default: false },
  flow_type: { type: String }
}, { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } })

if(mongoose.connection.models['newFlow'] !== undefined){
  NewFlowSchema = mongoose.model('newFlow');
}else{
  mongoose.model('newFlow', NewFlowSchema)
}
module.exports = ['newFlow', NewFlowSchema]
