// @builtioflow/models/oauths.js

const mongoose = require('mongoose')
const Schema = mongoose.Schema

const OauthSchema = new Schema({
  uid: { type: String },
  name: { type: String },
  version: { type: String },
  category: { type: String },
  user_uid: { type: String },
  requestor: { type: String, default: 'builtio' },
  published: { type: Boolean, default: false },
  title: { type: String },
  updated_at: { type: Date },
  deleted_at: { type: Date, default: null },
  icon: { type: String },
  created_at: { type: Date },
  express_enabled: { type: Boolean, default: false },
  advanced_enabled: { type: Boolean, default: true },
  development_phase: { type: String },
  desc: { type: String },
  oauth_type: { type: String },
  base_url: { type: String },
  env_local_params: [],
  env_development_params: [],
  env_staging_params: [],
  env_pre_prod_params: [],
  env_production_params: [],
  oauth2_no_state: { type: Boolean, default: false },
  oauth2_auth_url: { type: String },
  oauth2_auth_data: [],
  oauth2_auth_process: {},
  oauth2_auth_query: [],
  oauth2_token_url: { type: String },
  oauth2_token_method: { type: String },
  oauth2_token_process: {},
  oauth2_token_query: [],
  oauth2_token_data: [],
  oauth2_token_params: [],
  oauth2_token_headers: [],
  oauth2_refresh_url: { type: String },
  oauth2_refresh_query: [],
  oauth2_refresh_process: {},
  oauth2_refresh_data: [],
  oauth2_validate_url: { type: String },
  oauth2_validate_params: [],
  oauth2_validate_form: [],
  oauth2_validate_query: [],
  oauth2_validate_data: [],
  oauth2_validate_headers: [],
  oauth2_validate_method: { type: String },
  oauth2_params_scope: [],
  oauth2_params_other: [],
  oauth2_required: [],
  oauth2_request: {},
  oauth2_include: [],
  oauth2_resources: [],
  oauth1_request_url: { type: String },
  oauth1_request_format: { type: String },
  oauth1_auth_url: { type: String },
  oauth1_token_url: { type: String },
  oauth1_token_format: { type: String },
  oauth1_validate_url: { type: String },
  oauth1_validate_method: { type: String },
  oauth1_refresh: {},
  oauth1_auth_query: [],
  oauth1_validate_query: [],
  oauth1_validate_data: [],
  oauth1_validate_oauth: [],
  oauth1_validate_headers: [],
  oauth1_include: []
})

mongoose.model('Oauth', OauthSchema)
module.exports = ['Oauth', OauthSchema]
