// @builtioflow/models/plans.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const PlanSchema = new Schema({
    uid: { type: String },
    name: { type: String },
    plan_code: { type: String },
    description: { type: String },
    braintree_plan: { type: String },
    enterprise_plan: { type: Boolean },
    viewed: { type: Boolean },
    deleted_at: { type: Date, default: null },
    features: [{ type: Schema.Types.ObjectId, ref: 'Feature' }],
    features_json: { type: Schema.Types.Mixed },
    default: { type: Boolean },
    link1_url: { type: String },
    link1_text: { type: String },
    link2_url: { type: String },
    plan_type: { type: String, default: 'enterprise' },
    paid: { type: Boolean, default: false },
    link2_text: { type: String },
    required_plans: { type: Boolean },
    is_annual: { type: Boolean },
    is_monthly: { type: Boolean },
    plan_price: { type: String }
}, { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } });

mongoose.model('Plan', PlanSchema)
module.exports = ['Plan', PlanSchema]
