// @builtioflow/models/users.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const crypto = require('crypto');

const User = new Schema({
    uid: { type: String },
    plan: [{ type: Schema.Types.ObjectId, ref: 'Plan' }],
    company: { type: String, default: '' },
    locale: { type: String },
    phone: { type: String },
    fax: { type: String },
    merchantid: { type: String },
    credit_card: [{ "last_digits": Number, "card_token": String }],
    email: { type: String },
    first_name: { type: String, default: '' },
    last_name: { type: String, default: '' },
    job_title: { type: String },
    address: { type: String },
    password: { type: String },
    braintree_payment_method_token: { type: String },
    active: { type: Boolean },
    user_type: { type: String, default: 'enterprise' },
    account_type: { type: String, default: 'account' },
    type: { type: String, default: 'account' },
    force_suspend: { type: Boolean, default: false },
    subscription: {
        expired: { type: Boolean, default: false },
        days: { type: Number, default: 0 },
        start_date: { type: Date },
        hasUnlimitedPlan: { type: Boolean, default: false }
    },
    avatar: { type: String, default: "/img/user-img2.svg" },
    used_credits: { type: Number, default: 0 },
    allocated_credits: { type: Number, default: 0 },
    used_transactions: { type: Number, default: 0 },
    allocated_transactions: { type: Number, default: 0 },
    device_data: { type: Schema.Types.Mixed },
    device_type: { type: String },
    auth_data: [{
        "google": [{ "email": String, "user_profile": Schema.Types.Mixed }],
        "facebook": [{ "user_id": String, "user_profile": Schema.Types.Mixed }],
        "twitter": [{ "user_id": String, "user_profile": Schema.Types.Mixed }]
    }],
    last_login: { type: Number },
    activation_token: { type: String },
    password_token: { type: String },
    deleted_at: { type: Date, default: null },
    flows_created_count: { type: Number, default: 0 },
    built_uid: { type: String },
    flow_uid: { type: String },
    support_key: { type: String },
    first_login: { type: Boolean, default: true },
    activation_expired: { type: Boolean, default: false },
    auto_refresh: { type: Boolean, default: true },
    reminder_mail: { type: Boolean, default: false },
    braintree_customerId: { type: String },
    braintree_subscriptionId: { type: String },
    subscription_failure_count: { type: Number, default: 0 },
    metadata: { type: Schema.Types.Mixed },
    billing_start_date: { type: Date },
    billing_end_date: { type: Date },
    last_product_announcement_seen: { type: Date },
    set_disable_status: { type: Boolean, default: false },
    plans_limits: {
        isLightLimit: { type: Boolean, default: false },
        isStandardLimit: { type: Boolean, default: false },
        isPremiumLimit: { type: Boolean, default: false }
    },
    enable_twofa: { type: String, default: "disable" },
    authy_user_uid: { type: String },
    country_code: { type: String },
    authy_number: { type: String },
    payment_set: { type: String },
    failed_login_count: { type: Number, default: 0 },
    is_twofa_sms_enabled: { type: Boolean, default: false },
    last_password_hash: { type: Array },
    credit_limit_email: { type: Boolean, default: false },
    canvas_first_tour: { type: Boolean, default: false },
    partner: { type: Schema.Types.ObjectId, ref: 'Partner' }
}, { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } });

mongoose.model('User', User)
module.exports = ['User', User]
