// @builtioflow/models/userauths.js
var mongoose = require('mongoose');
var uuid = require("uuid/v4");
var Schema = mongoose.Schema;

var UserAuthSchema = new Schema({
    uid: { type: String },
    name: { type: String },
    provider_title: { type: String },
    category: { type: String },
    org_uid: { type: String },
    env_uid: { type: String },
    project_uid: { type: String },
    tenant_uid: { type: String, default: 'builtio' },
    data: { type: Schema.Types.Mixed },
    user: { type: String },
    description: { type: String },
    icon: { type: String },
    type: { type: String },
    auth: { type: String },
    created_at: { type: Date },
    updated_at: { type: Date },
    deleted_at: { type: Date, default: null },
    expires_at: { type: Number, default: 0 },
    provider_name: { type: String },
    provider_title: { type: String }
});

UserAuthSchema.pre('save', function (next) {
    var id = 'au' + uuid();
    this.created_at = this.updated_at = new Date();
    this.uid = id;
    this.data.code = id;
    next();
});

mongoose.model('UserAuth', UserAuthSchema);
module.exports = ['UserAuth', UserAuthSchema]
