// @builtioflow/models/userconnections.js
/**
 * Created by Pradeep on 12/23/15.
 */


var mongoose = require('mongoose');
var uuid = require('uuid/v4');
var Schema = mongoose.Schema;

var UserConnectionSchema = new Schema({
    uid: { type: String },
    version: { type: String },
    name: { type: String },
    label: { type: String },
    provider_title: { type: String },
    input: { type: Schema.Types.Mixed },
    updated_at: { type: Date },
    usage: { type: String },
    user: { type: String },
    data: { type: String },
    meta: { type: String },
    icon: { type: String },
    source: { type: String },
    app_id: { type: String },
    app_build_id: { type: String },
    app_version: { type: String },
    deleted_at: { type: Date, default: null },
    created_at: { type: Date, default: Date.now },
    org_uid: { type: String },
    env_uid: { type: String },
    project_uid: { type: String }
});

UserConnectionSchema.pre('save', function (next) {
    this.uid = 'conn' + uuid();
    this.created_at = this.updated_at = new Date();
    next();
});

mongoose.model('UserConnection', UserConnectionSchema);
module.exports = ['UserConnection', UserConnectionSchema]
