// @builtioflow/models/vqueues.js
const mongoose = require('mongoose')
const crypto = require('crypto')
const Schema = mongoose.Schema

const VQueue = new Schema({
  uid: { type: String },
  user: { type: String },
  flow_uid: { type: String },
  bill_uid: { type: String },
  name: { type: String },
  config: { type: String, default: 'default' },
  project_uid: { type: String },
  org_uid: { type: String, default: null },
  env_uid: { type: String, default: null },
  trigger: { type: String },
  payloadId: { type: String },
  sync: { type: Boolean, default: false },
  delaySeconds: { type: Number, default: 0 },
  headers: { type: String },
  reserved: { type: Number, default: 0 },
  runtime: { type: Number, default: 0 },
  queue: { type: Number },
  query: { type: String },
  method: { type: String },
  created_at: { type: Date, default: Date.now }
});

VQueue.pre('save', function (next) {
  if (!this || !this.uid) {
    this.uid = 'bid' + crypto.randomBytes(18).toString('hex')
  }
  next()
});

mongoose.model('VQueue', VQueue);
module.exports = ['VQueue', VQueue];
